File: leetcode/unique_length_3_palindromic_subsequences/python/solution.py
class Solution:
    def countPalindromicSubsequence(self, text: str) -> int:
        """
        Time complexity: O(n3)
        Auxiliary space complexity: O(1)
        Tags:
            DS: hash set
            A: iteration
        """
        palindrome_set = set()

        for index, letter in enumerate(text):
            for left in range(index):
                left_letter = text[left]

                for right in range(index + 1, len(text)):
                    right_letter = text[right]

                    if left_letter == right_letter:
                        palindrome_set.add(left_letter + letter + right_letter)

        return len(palindrome_set)


class Solution:
    def countPalindromicSubsequence(self, text: str) -> int:
        """
        Time complexity: O(n)
        Auxiliary space complexity: O(n)
        Tags:
            DS: hash map, hash set
            A: iteration
        """
        palindrome_set = set()
        left_letters = set()
        right_letters = {}

        for letter in text:
            right_letters[letter] = right_letters.get(letter, 0) + 1

        for middle_letter in text:
            right_letters[middle_letter] -= 1

            if right_letters[middle_letter] == 0:
                right_letters.pop(middle_letter)

            for i in range(26):
                letter = chr(ord("a") + i)

                if (
                    letter in left_letters and
                    letter in right_letters
                ):
                    palindrome_set.add(letter + middle_letter)

            left_letters.add(middle_letter)

        return len(palindrome_set)


print(Solution().countPalindromicSubsequence("aabca") == 3)
print(Solution().countPalindromicSubsequence("adc") == 0)
print(Solution().countPalindromicSubsequence("bbcbaba") == 4)
print(Solution().countPalindromicSubsequence("zqpppgacudvmqekmefkzyyfrffeylqrwxlupvskyonqsbclwwgnzbktzelwuhehxrxmqcnepxokialxxwciqsetcsqcsszpeobeiwwedtbisyhexyatammupmfrllpawhqvfebjdappicczehrsooztjatixvtvbmdwikffbozncspuslwgoqypmsmvwghfdmutfpkbjufqrgbhotcikoyvfvxmmadelwxmvybnoroapixubdvijnepeduiwshcwjvhnejafcnuxeimwiiucznzfakwdibwwixcttatqffhnurhecyocoohyuoeixobaxbjcksxqrljiftvcxtocusciqtmydxgjexiwimbcmvhjonkscobhlpggembfslzoisertsvcpiclikprpviqbfdptvtrlhqlfwhurxysxzppnwwbxzaozchalpqsklfedovjkhwdaqdxrzdduwxsyqllvkflamtshyoaamjpzcsnwthnnpgqrrroppxnalxoijzhesphugqporhtamdbugqhgtpxtrjeugenazzpvvtkjrsepvbgvbmmmyxgrkgnlhujinycnjvpqhhugplrgrunrziaabknrjsgaqbpxfpdycpjtquecehrblzurruguhbkzgvebzfkqcolpclgabsuamqaakdikasumksvbfjrudnzihbzqjwivthfozrhkcrmxleaazgkuqmzvzaiiskfrnywntgbtmaxqgqaqxvcpvbvcpqbfivtkdroizfbebhtejegpduqjewcaysphsumddhlgerpspcvhkoezzqwznmqfbcdvxmexbjfgqxlcbneanbglpktxfcfgkfxbpblfpejlfjhiaohcmktfheuyxpof") == 676)

File: leetcode/unique_length_3_palindromic_subsequences/javascript/solution.js
class Solution {
   /**
    * Time complexity: O(n)
    * Auxiliary space complexity: O(n)
    * Tags:
    *     DS: hash map, hash set
    *     A: iteration
    * @param {string} text
    * @return {number}
    */
   countPalindromicSubsequence(text) {
      const palindromeSet = new Set();
      const leftLetters = new Set();
      const rightLetters = new Map();

      for (const letter of text) {
         rightLetters.set(letter, (rightLetters.get(letter) || 0) + 1);
      }

      for (const middleLetter of text) {
         rightLetters.set(middleLetter, rightLetters.get(middleLetter) - 1);

         if (rightLetters.get(middleLetter) === 0) {
            rightLetters.delete(middleLetter);
         }

         for (let i = 0; i < 26; i++) {
            const letter = String.fromCharCode('a'.charCodeAt(0) + i);

            if (
               leftLetters.has(letter) &&
               rightLetters.has(letter)
            ) {
               palindromeSet.add(letter + middleLetter);
            }
         }

         leftLetters.add(middleLetter);
      }

      return palindromeSet.size;
   };
}


const countPalindromicSubsequence = new Solution().countPalindromicSubsequence;
console.log(new Solution().countPalindromicSubsequence('aabca') === 3)
console.log(new Solution().countPalindromicSubsequence('adc') === 0)
console.log(new Solution().countPalindromicSubsequence('bbcbaba') === 4)
console.log(new Solution().countPalindromicSubsequence('zqpppgacudvmqekmefkzyyfrffeylqrwxlupvskyonqsbclwwgnzbktzelwuhehxrxmqcnepxokialxxwciqsetcsqcsszpeobeiwwedtbisyhexyatammupmfrllpawhqvfebjdappicczehrsooztjatixvtvbmdwikffbozncspuslwgoqypmsmvwghfdmutfpkbjufqrgbhotcikoyvfvxmmadelwxmvybnoroapixubdvijnepeduiwshcwjvhnejafcnuxeimwiiucznzfakwdibwwixcttatqffhnurhecyocoohyuoeixobaxbjcksxqrljiftvcxtocusciqtmydxgjexiwimbcmvhjonkscobhlpggembfslzoisertsvcpiclikprpviqbfdptvtrlhqlfwhurxysxzppnwwbxzaozchalpqsklfedovjkhwdaqdxrzdduwxsyqllvkflamtshyoaamjpzcsnwthnnpgqrrroppxnalxoijzhesphugqporhtamdbugqhgtpxtrjeugenazzpvvtkjrsepvbgvbmmmyxgrkgnlhujinycnjvpqhhugplrgrunrziaabknrjsgaqbpxfpdycpjtquecehrblzurruguhbkzgvebzfkqcolpclgabsuamqaakdikasumksvbfjrudnzihbzqjwivthfozrhkcrmxleaazgkuqmzvzaiiskfrnywntgbtmaxqgqaqxvcpvbvcpqbfivtkdroizfbebhtejegpduqjewcaysphsumddhlgerpspcvhkoezzqwznmqfbcdvxmexbjfgqxlcbneanbglpktxfcfgkfxbpblfpejlfjhiaohcmktfheuyxpof') === 676)
